pkg/controller/job/job_controller.go

```diff
@@ -646,32 +724,276 @@ func (jm *Controller) syncJob(key string) (bool, error) {
+	if len(newUncounted) != len(uncountedStatus.Succeeded) {
+		needsFlush = true
+		job.Status.Succeeded += int32(len(uncountedStatus.Succeeded) - len(newUncounted))
+		uncountedStatus.Succeeded = newUncounted
+	}
+	newUncounted = updateUncountedWithFailedFinalizerRemovals(uncountedStatus.Failed, uidsWithFinalizer)
+	if len(newUncounted) != len(uncountedStatus.Failed) {
+		needsFlush = true
+		job.Status.Failed += int32(len(uncountedStatus.Failed) - len(newUncounted))
+		uncountedStatus.Failed = newUncounted
+	}
+	if needsFlush {
+		if err := jm.updateStatusHandler(job); err != nil {
+			return needsFlush, fmt.Errorf("removing uncounted pods from status: %w", err)
+		}
+	}
+	return needsFlush, rmErr
+}
+
+// removeTrackingFinalizerFromPods removes tracking finalizers from Pods and
+// returns the pod for which the operation failed (if the pod was deleted when
+// this function was called, it's considered as the finalizer was removed
+// successfully).
+func (jm *Controller) removeTrackingFinalizerFromPods(pods []*v1.Pod) ([]*v1.Pod, error) {
+	errCh := make(chan error, len(pods))
+	var failed []*v1.Pod
+	var lock sync.Mutex
+	wg := sync.WaitGroup{}
+	wg.Add(len(pods))
+	for i := range pods {
+		go func(i int) {
+			pod := pods[i]
+			defer wg.Done()
+			if patch := removeTrackingFinalizerPatch(pod); patch != nil {
+				if err := jm.podControl.PatchPod(pod.Namespace, pod.Name, patch); err != nil && !apierrors.IsNotFound(err) {
 					errCh <- err
 					utilruntime.HandleError(err)
+					lock.Lock()
+					failed = append(failed, pod)
+					lock.Unlock()
+					return
 				}
 			}
-		}(pods[i])
+		}(i)
 	}
 	wg.Wait()
-	return successfulDeletes, errorFromChannel(errCh)
+	return failed, errorFromChannel(errCh)
+}
+
+// enactJobFinished adds the Complete or Failed condition and records events.
+// Returns whether the Job was considered finished.
+func (jm *Controller) enactJobFinished(job *batch.Job, finishedCond *batch.JobCondition) bool {
+	if finishedCond == nil {
+		return false
+	}
+	if uncounted := job.Status.UncountedTerminatedPods; uncounted != nil {
+		if len(uncounted.Succeeded) > 0 || len(uncounted.Failed) > 0 {
+			return false
+		}
+	}
+	job.Status.Conditions = append(job.Status.Conditions, *finishedCond)
+	if finishedCond.Type == batch.JobComplete {
+		if job.Spec.Completions != nil && job.Status.Succeeded > *job.Spec.Completions {
+			jm.recorder.Event(job, v1.EventTypeWarning, "TooManySucceededPods", "Too many succeeded pods running after completion count reached")
+		}
+		job.Status.CompletionTime = &finishedCond.LastTransitionTime
+		jm.recorder.Event(job, v1.EventTypeNormal, "Completed", "Job completed")
+	} else {
+		jm.recorder.Event(job, v1.EventTypeWarning, finishedCond.Reason, finishedCond.Message)
+	}
+	return true
+}
+
+func updateUncountedWithFailedFinalizerRemovals(uncounted []types.UID, uidsWithFinalizer sets.String) []types.UID {
```

Comment:
nit: rename the function, we are not updating anything

Reply:
probably another leftover from a previous approach 😅